src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+  stack.push(_cl);
+  int idx = node_cnt - 1;
+  while (stack.length() > 0) {
+    Node* n = stack.top();
+    if (!visited.test(rpo_idx(n))) {
+      // Forward arc in graph
+      visited.set(rpo_idx(n));
+    } else if (!post_visited.test(rpo_idx(n))) {
+      // Cross or backward arc in graph
+      if (!is_memory_phi(n)) {
+        // Push all users in loop for non-mem-phi nodes
+        for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+          Node* use = n->fast_out(i);
+          if (in_body(use) && !visited.test(rpo_idx(use))) {
+            stack.push(use);
+          }
+        }
+      }
+      if (n == stack.top()) {
+        // Node is still at the top - no additional use is pushed, visit it.
+        // Also initialize node info at this time.
+        stack.pop();
+        assert(idx >= 0, "Is some node visited more than once?");
+        _body_nodes.at_put_grow(idx, n);
+        _elem_bt.at_put_grow(idx, T_ILLEGAL);
+        idx--;
+        post_visited.set(rpo_idx(n));
+      }
+    } else {
+      stack.pop();
+    }
+  }
+
+  // Bail out if loop has unreachable node while traversing from head
+  if (idx != -1) {
+    trace_msg(nullptr, "Loop has unreachable node while traversing from head");
+    return false;
+  }
+  // Create a real index map for future use
+  for (int i = 0; i < _body_nodes.length(); i++) {
+    set_rpo_idx(_body_nodes.at(i), i);
+  }
+
+#ifndef PRODUCT
+  if (TraceMaskedLoop) {
+    tty->print_cr("Collected loop nodes in reverse postorder");
+    for (int i = 0; i < _body_nodes.length(); i++) {
+      tty->print(" rpo=%d\t", i);
+      _body_nodes.at(i)->dump();
+    }
+    tty->cr();
+  }
+#endif
+
+  return true;
+}
+
+// Try including a node's input at specified index into current statement
+bool VectorMaskedLoop::collect_statements_helper(
+          Node* node, uint idx, Node_List* stmt, Node_List* worklist) {
+  Node* in = node->in(idx);
+  if (stmt->contains(in) || !in_body(in)) {
+    // Input is already included in current statement or out of loop
+    return true;
+  }
+
+  // Check support for special inputs first and then general ones
+  if (is_loop_iv_or_incr(in)) {
+    // 1) Check the support of loop iv or increment node input
+    BasicType bt = statement_bottom_type(stmt);
+    bt = is_subword_type(bt) ? bt : T_INT;
+    if (VectorNode::is_populate_index_supported(bt)) {
+      return true;
+    } else {
+      trace_msg(in, "Populate index operation is not supported");
+      return false;
+    }
+  } else if (in->is_Phi()) {
+    // 2) We don't support phi nodes except the iv phi of the loop
+    trace_msg(in, "Found unsupported phi input");
+    return false;
+  } else if (in->is_Load()) {
+    // 3) Ok to include a load node if it's supported memory access
+    if (supported_mem_access(in->as_Load())) {
+      stmt->push(in);
+      return true;
+    } else {
+      trace_msg(in, "Found unsupported memory load input");
```

Comment:
This is a bit generic. Would be nice to have more specific info why it is "unsupported". See my example that hit it.

Reply:
Good suggestion! I have added more `trace_msg()` calls in `VectorMaskedLoop::mem_access_to_swpointer`.